### ISSUE TITLE
Parsable `hq jobs` output

### ISSUE THREAD

**Issue by giovannipizzi** [It4innovations/hyperqueue#241]
Hi, thanks for this very nice package! It actually covers exactly one of our usecases (from what we understand from some preliminary testing) and we plan to support it as one of the schedulers in our AiiDA high-throughput platform (see e.g. [this GitHub discussion thread](https://github.com/aiidateam/aiida-core/discussions/5112#discussioncomment-1780979)) by implementing a plugin for it.

I'd have some preliminary feedback/suggestions that would make writing such a plugin much easier, I'll create a few issues.

This is about the possibility of having a flag, e.g. `--parsable` to `hq jobs`, that outputs the job status etc. not in human-readable form (like a ASCII table) but in some other format easy to parse (e.g. YAML, but any other format would also do I think, like CSV [but one has to be careful to escape commas if they appear e.g. in the job name, or wrap in double quotes)). Would this be feasible? Also, maybe one could have a command line option to specify which columns to output?

Also (partially unrelated): the default of `hq jobs` is to show *all* jobs, and your can specify the list of statuses to filter. I think the in most schedulers it's common to show only "active" jobs, I think equivalent to `hq jobs waiting running`. Would it be possible to get this output easily? E.g. I'm thinking of a few options:
- make `hq jobs` return only active jobs, have an `--all` option to show all (as you do e.g. for workers and allocs)
- keep the current behaviour have a flag `--active` to show only active states - this is mostly in case in the future you introduce more "active" = "non-terminal" states, so the client does not need to know the explicit list (e.g. if you introduce a "on hold" status or similar)

**Comment by giovannipizzi** [comment]
A question that is partially related is what is: is the output format of `hq` expected to be stable over time, or you plan to change the output? This because we're planning to write wrappers to parse its output (of `hq jobs`, `hq submit` (to get the Job ID), `hq cancel` (to check if it was successful), `hq job` (to get detailed job info after the job is over, similar to `sacct` in SLURM), ...) so I'd like to know if we can assume the output format is stable or we should have the plugin detect which version is running and support multiple versions

**Comment by Kobzol** [comment]
Hi! Your suggestion regarding `hq jobs` sounds reasonable, I created a separate issue for that (https://github.com/It4innovations/hyperqueue/issues/245) so that this one only focuses on parseable output.

HQ actually already supports machine-readable (JSON) output for most commands (https://github.com/It4innovations/hyperqueue/issues/28). You can enable it using the `--output-type=json` flag after `hq`, e.g.:
```bash
$ hq --output-type=json jobs
```

However, the JSON output format is currently not documented nor stabilized, because we didn't have any users for it yet, so right now it outputs semi-internal data. Once we flesh out the information that would actually be useful for the consumers of this output, we intend to stabilize it and keep it backwards compatible to a reasonable degree.

It would be useful to us if you can tell us all the commands that you intend to use in a machine-readable way, what kind of data do you want to extract out of them and what format would be reasonable for the individual fields (e.g. we can produce dates/times in ISO format etc.).

**Comment by giovannipizzi** [comment]
Ok, thanks a lot! I didn't know about the output-type flag - actually I tried and in my version only `cli` is enabled. I imagine this is a recent addition (actually I see you are working quite actively on this, cool!)

Regarding machine-readable parsing: at the moment I think we query only:

- `hq jobs` to know the state of submitted jobs
- the output of `hq submit` to get if it was successful (can we rely not he return code being zero or not?) and to parse the JOB ID
- `hq cancel` to get if this was successful (can we rely not he return code being zero or not? Also, probably the fact the job was already canceled or already finished should not be an error, in the way we envision it: if the end goal is to avoid it to continue, the result is "OK", while an error would be e.g. "I couldn't communicate with the server so I didn't manage to cancel it" - if you prefer to make also double canceling and error, it would be good to use a range on return codes that signals non-critical errors, so we can just check the error code and consider the run OK if the value in in that range).

In addition, we get (at the end of every job) a detailed information on the just-finished job (e.g. with sacct on SLURM), because typically this information is deleted after a few days in our experience with a few schedulers. Having this parsable would also be great for later debugging, but we don't rely on this to be parsable for standard usage (we just get the output and store it).

The point about times is a good one. Indeed, in SLURM we now run the `squeue` command as this to have an easy-to-parse format and not depending on the locale (I tried to reconstruct it, I might be wrong): 
```
"SLURM_TIME_FORMAT='standard'" 'squeue' '--noheader' "-o '%i,%t,%r,%B,%u,%D,%C,%R,%P,%l,%M,%S,%j,%V'"
```
I think that the best about the time format is to use the ISO format (at least in local computer time; if possible also the timezone, but I know this is trickier :-) ).

**Comment by Kobzol** [comment]
Thank you for the detailed description. You can try either the `quiet` or `json` output modes to see what suits you better. I will go through the JSON output and try to consolidate it, create some stable description and add it to the documentation.

Btw, we are also in the process of making Python bindings for HQ, would that be useful for you?

**Comment by giovannipizzi** [comment]
Indeed I saw the python binding issue! 

Not directly - even if AiiDA is in python, it lives on an external machine and connects to the supercomputer, so it calls any submission as an external executable. 

But I think the python bindings can be very useful in a number of other cases!

**Comment by giovannipizzi** [comment]
A few comments/suggestions on things I would suggest to change in the current JSON, for your ref when you'll go through it, that I found while trying to get the data to show when and where jobs were being run (a sort of hand-made dashboard).

- output all times in UTC. Most are, but for instance workers instead output the time since the epoch split in two parts and need to be reconstructed like this: `status['started_at']['secs_since_epoch'] + status['started_at']['nanos_since_epoch'] * 1.e-9)` (plus converting epoch to date time
- (if possible, but not critical) it would be good to output somewhere (e.g. in the output of `hq server info`?) the timezone of the computer, if it's easy to get - e.g. I wanted to compare the time printed by the code with the time of the scheduler (e.g. to check if the job was starting right away or being blocked by SLURM policies), knowing the timezone would help - but if you don't implement it it's ok)
- the output of `hq --output-type=json job` prints two lines, a JSON each (and I'm not sure what it is), or multiple lines if more jobs are requested. I suggest to put everything in a single JSON, possibly e.g. with `{'jobs': [...]}` where `[...]` is a list of job information, all in the same JSON
- I find a bit strange that to access, e.g. the start date or end date, I need to go to, e.g. `job_data['job_detail']['tasks'][0]['state']['Finished']['start_date']`, i.e. in particular that `Finished` is a key. I would rather have the task information (as now) as `task_data = job_data['job_detail']['tasks'][0]` and within it e.g. `{'state': 'Finished', 'start_date': .., 'end_date', ...}`
- similarly, to get the number of CPUs, I have to do `job_data['job_detail']['info']['resources']['cpus']['Compact']`; I would suggest to have something like `job_data['job_detail']['info']['resources']['cpus'] = {'mode': 'Compact', 'number': 16, ...}`
- in the output of `hq worker list`, for each worker, `worker['configuration']['extra']['JobManager']` is not a dictionary but a string that encodes a JSON with additional data, so one would need to get the JSON, then get this string and decode with JSON again - I suggest to just store the dictionary and not a JSON-encoded string

**Comment by Kobzol** [comment]
Thanks for the description. As you have probably noticed, the current JSON output is not manually designed, we just take internal structures and dump them to JSON. This creates really weird things like `nanos_since_epoch`, JSON within JSON, weird keys etc.

I'll try to design a more reasonable JSON output that will be manually crafted.

Regarding time, I'll try to use UTC as much as possible, however there might be some issues with time in general, as it happens with every distributed system. Some of the times come from different nodes, so they might not be fully synchronized. But on a reasonable cluster it should be OK, if maximum accuracy is not needed.

**Comment by Kobzol** [comment]
@giovannipizzi JSON output has been reworked and should now be much easier to work with. You can now examples of JSON output for various commands [here](https://it4innovations.github.io/hyperqueue/latest/cli/output-mode/#json). I didn't create a full (JSON) schema yet, but I hope that this should improve your experience with the JSON output mode.
